src/DataProtection/DataProtection/src/TypeForwardingActivator.cs

```
@@ -64,6 +62,27 @@ internal object CreateInstance(Type expectedBaseType, string originalTypeName, o
         }
 
         protected string RemoveVersionFromAssemblyName(string forwardedTypeName)
-            => _versionPattern.Replace(forwardedTypeName, "");
+        {
+            // Type, Assembly, Version={Version}, Culture={Culture}, PublicKeyToken={Token}
+
+            var versionStartIndex = forwardedTypeName.IndexOf("Version=", StringComparison.Ordinal);
+            while (versionStartIndex != -1)
+            {
+                var versionEndIndex = forwardedTypeName.IndexOf(',', versionStartIndex + "Version=".Length);
+
+                if (versionEndIndex == -1)
+                {
+                    // No end index?
+                    return forwardedTypeName;
```

Shouldn't this return `forwardedTypeName.Substring(0, versionStartIndex);` or something like that. e.g. `Foo, MyAssembly, Verison=3.1`

---

I'm not sure we have any tests or we'd ever expect to hit this condition, should we just throw instead?  Its basically invalid data from what expect which is the normal full type string with culture and public key

---

Yup. Silently falling upwards feels weird

---

Updated to throw InvalidOperationException

---

Oops I guess I was wrong, that case does exist and has tests:

```
            var name = "Microsoft.AspNet.DataProtection.TypeForwardingActivatorTests+ClassWithParameterlessCtor, Microsoft.AspNet.DataProtection.Tests, Version=1.0.0.0";
```

So I guess i'm going to revert this back to the original return, and remove the comment

---

Okay tweaked the code to look for the ", Version=", and if no comma is found, it just drops everything after.